plugins/saveactions.c

```diff
@@ -304,6 +314,274 @@ static void instantsave_document_new_cb(GObject *obj, GeanyDocument *doc, gpoint
+		{
+			dialogs_show_msgbox(GTK_MESSAGE_ERROR,
+				_("Persistent temp files directory does not exist or is not writable."));
+			return;
+		}
+
+		new_temp_file_name_utf8 = create_new_temp_file_name();
+
+		if (new_temp_file_name_utf8 == NULL)
+		{
+			return;
+		}
+
+		temp_files_dir_utf8 = utils_get_utf8_from_locale(persistent_temp_files_target_dir);
+
+		new_temp_file_path_utf8 = g_strconcat(temp_files_dir_utf8,
+			G_DIR_SEPARATOR_S, new_temp_file_name_utf8, NULL);
+
+		document_save_file_as(doc, new_temp_file_path_utf8);
+
+		g_free(new_temp_file_name_utf8);
+		g_free(temp_files_dir_utf8);
+		g_free(new_temp_file_path_utf8);
+	}
+}
+
+
+static gboolean document_is_empty(GeanyDocument *doc)
+{
+	/* if total text length is 1 while line count is 2 - then the only character of whole text is a linebreak,
+	which is how completely empty document saved by Geany to disk looks like */
+	return sci_get_length(doc->editor->sci) == 0
+		|| (sci_get_length(doc->editor->sci) == 1 && sci_get_line_count(doc->editor->sci) == 2);
+}
+
+
+/* Open document in idle - useful when trying to re-open document from 'document-close' callback
+ *
+ * @param pointer to document path (locale-encoded)
+ *
+ * @return always FALSE = Just a one shot execution
+ *
+ */
+static gboolean open_document_once_idle(gpointer p_locale_file_path)
+{
+	gchar *locale_file_path = (gchar *) p_locale_file_path;
+
+	document_open_file(locale_file_path, FALSE, NULL, NULL);
+
+	g_free(locale_file_path);
+
+	return FALSE;
+}
+
+
+static gint run_unsaved_dialog_for_persistent_temp_files_tab_closing(const gchar *msg, const gchar *msg2)
+{
+	GtkWidget *dialog, *button;
+	gint ret;
+
+	dialog = gtk_message_dialog_new(GTK_WINDOW(geany->main_widgets->window), GTK_DIALOG_DESTROY_WITH_PARENT,
+			GTK_MESSAGE_QUESTION, GTK_BUTTONS_NONE, "%s", msg);
+	gtk_window_set_title(GTK_WINDOW(dialog), _("Question"));
+	gtk_message_dialog_format_secondary_text(GTK_MESSAGE_DIALOG(dialog), "%s", msg2);
+	gtk_dialog_add_button(GTK_DIALOG(dialog), GTK_STOCK_CANCEL, GTK_RESPONSE_CANCEL);
+
+	button = ui_button_new_with_image(GTK_STOCK_CLEAR, _("_Don't save"));
+	gtk_dialog_add_action_widget(GTK_DIALOG(dialog), button, GTK_RESPONSE_NO);
+	gtk_widget_show(button);
+
+	gtk_dialog_add_button(GTK_DIALOG(dialog), GTK_STOCK_SAVE, GTK_RESPONSE_YES);
+
+	gtk_dialog_set_default_response(GTK_DIALOG(dialog), GTK_RESPONSE_YES);
+	ret = gtk_dialog_run(GTK_DIALOG(dialog));
+
+	gtk_widget_destroy(dialog);
+
+	return ret;
+}
+
+
+static void show_unsaved_dialog_for_persistent_temp_files_tab_closing(
+	GeanyDocument *doc, const gchar *short_filename)
+{
+	gchar *msg, *old_file_path, *locale_file_path;
+	const gchar *msg2;
+	gint response;
+
+	msg = g_strdup_printf(_("The file '%s' is not saved."), short_filename);
+	msg2 = _("Do you want to save it before closing?");
+
+	response = run_unsaved_dialog_for_persistent_temp_files_tab_closing(msg, msg2);
+	g_free(msg);
+
+	locale_file_path = g_strdup(doc->real_path);
```

Comment:
strdup is needed because path is changed after save actions is performed below. 
But i lowered scope of that variable and used real_path where possible

Reply:
You're right, I missed that.